File: apps/api/internal/shooter/specials.go
package shooter

func (sim *simulation) activateRescue() {
	if sim.rescueCharge < RescueChargeLimit || sim.health <= 0 {
		return
	}
	sim.rescueCharge = 0
	sim.rescuesUsed++
	sim.lastRescueTick = sim.tick
	damage := sim.runtime.rescueDamage
	sim.shield += sim.runtime.guardOnSpecial
	behavior := sim.config.Kit.SpecialBehavior
	if behavior == "" {
		behavior = map[KitID]string{KitNana: "barrage_break", KitJiaran: "cheer_guard", KitXiangwan: "afterimage_replay", KitBella: "captain_parry", KitLulu: "subtitle_flip", KitXingtong: "prism_shift", KitNailu: "memory_bloom"}[sim.config.Kit.ID]
	}
	switch behavior {
	case "barrage_break":
		damage += sim.combo * 2
		for index := range sim.enemies {
			if sim.enemies[index].marks == 0 {
				continue
			}
			extra := sim.enemies[index].marks * max(4, damage/3)
			sim.enemies[index].health -= extra
			sim.addEffect("mark_detonation", sim.enemies[index].x, sim.enemies[index].y, 24, extra)
			sim.enemies[index].marks = 0
		}
	case "cheer_guard":
		sim.shield += 8
	case "afterimage_replay":
		damage += len(sim.enemyProjectiles) / 3
		for _, offset := range []int{-180, 0, 180} {
			if len(sim.playerProjectiles) >= sim.config.Limits.PlayerProjectiles {
				break
			}
			sim.nextProjectileID++
			sim.playerProjectiles = append(sim.playerProjectiles, projectileEntity{id: sim.nextProjectileID, x: clamp(sim.playerX+offset, playerRadius, ArenaWidth-playerRadius), y: playerY + 240, vy: -205, damage: max(1, sim.runtime.damage)})
		}
		sim.addEffect("afterimage_replay", sim.playerX, playerY, 36, damage)
	case "captain_parry":
		sim.shield += 18
		for _, vx := range []int{-140, -70, 0, 70, 140} {
			if len(sim.playerProjectiles) >= sim.config.Limits.PlayerProjectiles {
				break
			}
			sim.nextProjectileID++
			sim.playerProjectiles = append(sim.playerProjectiles, projectileEntity{id: sim.nextProjectileID, x: sim.playerX, y: playerY, vx: vx, vy: -190, damage: max(1, sim.runtime.damage)})
		}
		sim.addEffect("captain_parry", sim.playerX, playerY, 24, sim.shield)
	case "subtitle_flip":
		damage += len(sim.enemyProjectiles) / 2
		for len(sim.enemyProjectiles) > 0 && len(sim.playerProjectiles) < sim.config.Limits.PlayerProjectiles {
			bullet := sim.enemyProjectiles[len(sim.enemyProjectiles)-1]
			sim.enemyProjectiles = sim.enemyProjectiles[:len(sim.enemyProjectiles)-1]
			sim.nextProjectileID++
			sim.playerProjectiles = append(sim.playerProjectiles, projectileEntity{id: sim.nextProjectileID, x: bullet.x, y: bullet.y, vy: -190, damage: max(1, sim.runtime.damage/2), pierce: 1})
		}
		sim.addEffect("subtitle_flip", sim.playerX, playerY, 30, damage)
	case "prism_shift":
		damage += max(4, sim.runtime.damage)
		for index := range sim.enemies {
			if abs(sim.enemies[index].x-sim.playerX) <= 420 {
				sim.enemies[index].health -= damage
			}
		}
		sim.addEffect("prism_shift", sim.playerX, 0, 30, damage)
	case "memory_bloom":
		plants := sim.effects[:0]
		bloomed := 0
		for _, effect := range sim.effects {
			if effect.kind != "memory_plant" {
				plants = append(plants, effect)
				continue
			}
			bloomed++
			for index := range sim.enemies {
				if square(sim.enemies[index].x-effect.x)+square(sim.enemies[index].y-effect.y) <= square(720) {
					sim.enemies[index].health -= max(4, effect.power*2)
				}
			}
		}
		sim.effects = plants
		if bloomed > 0 {
			sim.health = min(sim.runtime.maxHealth, sim.health+1)
			sim.shield += bloomed
		}
		sim.addEffect("memory_bloom", sim.playerX, playerY, 45, bloomed)
	}
	for index := range sim.enemies {
		sim.enemies[index].health -= damage
	}
	sim.enemyProjectiles = sim.enemyProjectiles[:0]
	sim.combo += 3
	sim.comboClock = 120 + sim.runtime.comboExtend
	sim.score += 250
}

func (sim *simulation) earnRescue(amount int) {
	if amount <= 0 {
		return
	}
	// Encore content authors express this value as a penalty: positive values
	// must make Rescue slower, never faster. Keep at least one point for a
	// legitimate scoring event so low-value pickups still provide feedback.
	multiplier := 100 - sim.config.SpecialChargePenaltyPercent
	amount = max(1, amount*multiplier/100)
	sim.rescueCharge = min(RescueChargeLimit, sim.rescueCharge+amount)
}

File: apps/miniapp/src/features/shooter/specials.ts
import {
  PLAYER_RADIUS,
  PLAYER_Y,
  SHOOTER_WIDTH,
  clamp,
  goDivide,
  squaredDistance,
} from "@/features/shooter/constants";
import type { ShooterMutableState } from "@/features/shooter/types";
import { addPlayerProjectile, addShooterEffect } from "@/features/shooter/weapons";

export const earnRescue = (state: ShooterMutableState, amount: number): void => {
  if (amount <= 0) return;
  const adjusted = Math.max(
    1,
    goDivide(
      amount * (100 - state.config.special_charge_penalty_percent),
      100,
    ),
  );
  state.rescueCharge = Math.min(100, state.rescueCharge + adjusted);
};

const defaultBehavior: Record<string, string> = {
  nana7mi: "barrage_break",
  jiaran: "cheer_guard",
  xiangwan: "afterimage_replay",
  bella: "captain_parry",
  lulu: "subtitle_flip",
  xingtong: "prism_shift",
  nailu: "memory_bloom",
};

export const activateRescue = (state: ShooterMutableState): boolean => {
  if (state.rescueCharge < 100 || state.health <= 0) return false;
  state.rescueCharge = 0;
  state.rescuesUsed += 1;
  state.lastRescueTick = state.tick;
  let damage = state.runtime.rescueDamage;
  state.shield += state.runtime.guardOnSpecial;
  const behavior = state.config.kit.special_behavior || defaultBehavior[state.config.kit.id];
  if (behavior === "barrage_break") {
    damage += state.combo * 2;
    for (const enemy of state.enemies) {
      if (enemy.marks === 0) continue;
      const extra = enemy.marks * Math.max(4, goDivide(damage, 3));
      enemy.health -= extra;
      addShooterEffect(state, "mark_detonation", enemy.x, enemy.y, 24, extra);
      enemy.marks = 0;
    }
  } else if (behavior === "cheer_guard") state.shield += 8;
  else if (behavior === "afterimage_replay") {
    damage += goDivide(state.enemyProjectiles.length, 3);
    for (const offset of [-180, 0, 180]) {
      if (!addPlayerProjectile(state, { x: clamp(state.playerX + offset, PLAYER_RADIUS, SHOOTER_WIDTH - PLAYER_RADIUS), y: PLAYER_Y + 240, vy: -205, damage: Math.max(1, state.runtime.damage) })) break;
    }
    addShooterEffect(state, "afterimage_replay", state.playerX, PLAYER_Y, 36, damage);
  } else if (behavior === "captain_parry") {
    state.shield += 18;
    for (const vx of [-140, -70, 0, 70, 140]) {
      if (!addPlayerProjectile(state, { x: state.playerX, y: PLAYER_Y, vx, vy: -190, damage: Math.max(1, state.runtime.damage) })) break;
    }
    addShooterEffect(state, "captain_parry", state.playerX, PLAYER_Y, 24, state.shield);
  } else if (behavior === "subtitle_flip") {
    damage += goDivide(state.enemyProjectiles.length, 2);
    while (state.enemyProjectiles.length > 0 && state.playerProjectiles.length < state.config.limits.player_projectiles) {
      const bullet = state.enemyProjectiles.pop()!;
      addPlayerProjectile(state, { x: bullet.x, y: bullet.y, vy: -190, damage: Math.max(1, goDivide(state.runtime.damage, 2)), pierce: 1 });
    }
    addShooterEffect(state, "subtitle_flip", state.playerX, PLAYER_Y, 30, damage);
  } else if (behavior === "prism_shift") {
    damage += Math.max(4, state.runtime.damage);
    for (const enemy of state.enemies) if (Math.abs(enemy.x - state.playerX) <= 420) enemy.health -= damage;
    addShooterEffect(state, "prism_shift", state.playerX, 0, 30, damage);
  } else if (behavior === "memory_bloom") {
    const plants = [];
    let bloomed = 0;
    for (const effect of state.effects) {
      if (effect.kind !== "memory_plant") {
        plants.push(effect);
        continue;
      }
      bloomed += 1;
      for (const enemy of state.enemies) {
        if (squaredDistance(enemy.x, enemy.y, effect.x, effect.y) <= 720 ** 2) enemy.health -= Math.max(4, effect.power * 2);
      }
    }
    state.effects = plants;
    if (bloomed > 0) {
      state.health = Math.min(state.runtime.maxHealth, state.health + 1);
      state.shield += bloomed;
    }
    addShooterEffect(state, "memory_bloom", state.playerX, PLAYER_Y, 45, bloomed);
  }
  for (const enemy of state.enemies) enemy.health -= damage;
  state.enemyProjectiles = [];
  state.combo += 3;
  state.comboClock = 120 + state.runtime.comboExtend;
  state.score += 250;
  return true;
};
